salt/states/git.py

```diff
@@ -1588,6 +1588,455 @@ def present(name,
+                        )
+                    )
+            if not os.path.isabs(ident_path):
+                return _fail(
+                    ret,
+                    'Identity \'{0}\' is not an absolute path'.format(
+                        ident_path
+                    )
+                )
+    if https_user is not None and not isinstance(https_user, six.string_types):
+        https_user = str(https_user)
+    if https_pass is not None and not isinstance(https_pass, six.string_types):
+        https_pass = str(https_pass)
+
+    if os.path.isfile(target):
+        return _fail(
+            ret,
+            'Target \'{0}\' exists and is a regular file, cannot proceed'
+            .format(target)
+        )
+
+    try:
+        desired_fetch_url = salt.utils.url.add_http_basic_auth(
+            name,
+            https_user,
+            https_pass,
+            https_only=True
+        )
+    except ValueError as exc:
+        return _fail(ret, exc.__str__())
+
+    redacted_fetch_url = salt.utils.url.redact_http_basic_auth(desired_fetch_url)
+
+    # Check if onlyif or unless conditions match
+    run_check_cmd_kwargs = {'runas': user}
+    if 'shell' in __grains__:
+        run_check_cmd_kwargs['shell'] = __grains__['shell']
+    cret = mod_run_check(
+        run_check_cmd_kwargs, onlyif, unless
+    )
+    if isinstance(cret, dict):
+        ret.update(cret)
+        return ret
+
+    # Determine if supplied ref is a hash
+    remote_ref_type = 'ref'
+    if len(ref) <= 40 \
+            and all(x in string.hexdigits for x in ref):
+        ref = ref.lower()
+        remote_ref_type = 'hash'
+
+    comments = []
+    hash_exists_locally = False
+    local_commit_id = None
+
+    gitdir = os.path.join(target, '.git')
+    if os.path.isdir(gitdir) or __salt__['git.is_worktree'](target):
+        # Target directory is a git repository or git worktree
+
+        local_commit_id = _get_local_rev_and_branch(target, user)[0]
+
+        if remote_ref_type is 'hash' and __salt__['git.describe'](ref):
+            # The ref is a hash and it exists locally so skip to checkout
+            hash_exists_locally = True
+        else:
+            # Check that remote is present and set to correct url
+            remotes = __salt__['git.remotes'](target,
+                                              user=user,
+                                              redact_auth=False)
+
+            if remote in remotes and name in remotes[remote]['fetch']:
+                pass
+            else:
+                # The fetch_url for the desired remote does not match the
+                # specified URL (or the remote does not exist), so set the
+                # remote URL.
+                current_fetch_url = None
+                if remote in remotes:
+                    current_fetch_url = remotes[remote]['fetch']
+
+                if __opts__['test']:
+                    ret['changes']['remote'] = {'old': str(current_fetch_url), 'new': name}
```

Comment:
Can you remove this line? I recently changed the logic in git.latest to remove minor changes from the changes dictionary, because content in the changes dict will trigger watch requisites, and we only want to trigger these when the content managed by the state changes (i.e. when HEAD is reset, or the repo is cloned, etc.).